File: sdks/remote_plugin/python/voce/service/converters.py
from __future__ import annotations

import json
from typing import Any

from voce.core import (
    Field,
    MultiTrackConfig,
    PluginMetadata,
    PortMetadata,
    Property,
    TrackConfig,
)
from voce.proto import plugin_pb2 as pb

EVENT_TYPES = {
    "signal": "EVENT_TYPE_SIGNAL",
    "payload": "EVENT_TYPE_PAYLOAD",
    "audio": "EVENT_TYPE_AUDIO",
    "video": "EVENT_TYPE_VIDEO",
}

VALUE_TYPES = {
    "string": "VALUE_TYPE_STRING",
    "number": "VALUE_TYPE_NUMBER",
    "float": "VALUE_TYPE_NUMBER",
    "integer": "VALUE_TYPE_INTEGER",
    "int": "VALUE_TYPE_INTEGER",
    "boolean": "VALUE_TYPE_BOOLEAN",
    "bool": "VALUE_TYPE_BOOLEAN",
    "object": "VALUE_TYPE_OBJECT",
    "array": "VALUE_TYPE_ARRAY",
}

DROP_STRATEGIES = {
    "block_if_full": "DROP_STRATEGY_BLOCK_IF_FULL",
    "block": "DROP_STRATEGY_BLOCK_IF_FULL",
    "drop_newest": "DROP_STRATEGY_DROP_NEWEST",
    "drop_oldest": "DROP_STRATEGY_DROP_OLDEST",
}


def plugin_metadata_to_proto(metadata: PluginMetadata) -> pb.PluginMetadata:
    result = pb.PluginMetadata(
        name=metadata.name,
        description=metadata.description,
        schema=json_string(metadata.schema or {}),
        inputs=[property_to_proto(item) for item in metadata.inputs],
        outputs=[property_to_proto(item) for item in metadata.outputs],
        ports=[port_metadata_to_proto(item) for item in metadata.ports],
    )
    if metadata.multi_track is not None:
        result.multi_track.CopyFrom(multi_track_config_to_proto(metadata.multi_track))
    return result


def multi_track_config_to_proto(config: MultiTrackConfig) -> pb.MultiTrackConfig:
    result = pb.MultiTrackConfig(enabled=config.enabled)
    if config.payload is not None:
        result.payload.CopyFrom(track_config_to_proto(config.payload))
    return result


def track_config_to_proto(config: TrackConfig) -> pb.TrackConfig:
    return pb.TrackConfig(
        enabled=config.enabled,
        buffer_size=config.buffer_size,
        drop_strategy=drop_strategy_to_proto(config.drop_strategy),
        interrupt_signals=list(config.interrupt_signals),
    )


def property_to_proto(prop: Property) -> pb.Property:
    return pb.Property(
        type=event_type_to_proto(prop.type),
        name=prop.name,
        fields=[field_to_proto(item) for item in prop.fields],
    )


def field_to_proto(field: Field) -> pb.Field:
    return pb.Field(
        key=field.key,
        type=value_type_to_proto(field.type),
        required=field.required,
    )


def port_metadata_to_proto(port: PortMetadata) -> pb.PortMetadata:
    return pb.PortMetadata(
        type=event_type_to_proto(port.type),
        port=port.port,
        name=port.name,
        description=port.description,
    )


def event_type_to_proto(value: str) -> str:
    return EVENT_TYPES.get(value.lower(), "EVENT_TYPE_UNSPECIFIED")


def value_type_to_proto(value: str) -> str:
    return VALUE_TYPES.get(value.lower(), "VALUE_TYPE_UNSPECIFIED")


def drop_strategy_to_proto(value: str) -> str:
    return DROP_STRATEGIES.get(value.lower(), "DROP_STRATEGY_UNSPECIFIED")


def json_bytes(value: dict[str, Any]) -> bytes:
    return json.dumps(value, ensure_ascii=False).encode()


def json_string(value: dict[str, Any]) -> str:
    return json.dumps(value, ensure_ascii=False)


def json_from_bytes(data: bytes) -> dict[str, Any]:
    if not data:
        return {}
    value = json.loads(data.decode())
    if not isinstance(value, dict):
        raise ValueError("properties must decode to a JSON object")
    return value

File: sdks/remote_plugin/nodejs/src/internal/converters.ts
import type {
  PluginMetadata as SdkPluginMetadata,
  PropertyDef,
  FieldDef,
  PortMetadata as SdkPortMetadata,
  MultiTrackConfig as SdkMultiTrackConfig,
  TrackConfig as SdkTrackConfig,
} from '@/plugin.js'
import {
  DropStrategy,
  EventType,
  ValueType,
  type PluginMetadata as ProtoPluginMetadata,
  type Property as ProtoProperty,
  type Field as ProtoField,
  type PortMetadata as ProtoPortMetadata,
  type MultiTrackConfig as ProtoMultiTrackConfig,
  type TrackConfig as ProtoTrackConfig,
} from '@/proto/plugin.js'

// ---------------------------------------------------------------------------
// Event type mapping
// ---------------------------------------------------------------------------

const EVENT_TYPE_MAP: Record<string, EventType> = {
  signal: EventType.EVENT_TYPE_SIGNAL,
  payload: EventType.EVENT_TYPE_PAYLOAD,
  audio: EventType.EVENT_TYPE_AUDIO,
  video: EventType.EVENT_TYPE_VIDEO,
}

function eventTypeToProto(value: string): EventType {
  return EVENT_TYPE_MAP[value.toLowerCase()] ?? EventType.EVENT_TYPE_UNSPECIFIED
}

// ---------------------------------------------------------------------------
// Value type mapping
// ---------------------------------------------------------------------------

const VALUE_TYPE_MAP: Record<string, ValueType> = {
  string: ValueType.VALUE_TYPE_STRING,
  number: ValueType.VALUE_TYPE_NUMBER,
  float: ValueType.VALUE_TYPE_NUMBER,
  integer: ValueType.VALUE_TYPE_INTEGER,
  int: ValueType.VALUE_TYPE_INTEGER,
  boolean: ValueType.VALUE_TYPE_BOOLEAN,
  bool: ValueType.VALUE_TYPE_BOOLEAN,
  object: ValueType.VALUE_TYPE_OBJECT,
  array: ValueType.VALUE_TYPE_ARRAY,
}

function valueTypeToProto(value: string): ValueType {
  return VALUE_TYPE_MAP[value.toLowerCase()] ?? ValueType.VALUE_TYPE_UNSPECIFIED
}

// ---------------------------------------------------------------------------
// Drop strategy mapping
// ---------------------------------------------------------------------------

const DROP_STRATEGY_MAP: Record<string, DropStrategy> = {
  block_if_full: DropStrategy.DROP_STRATEGY_BLOCK_IF_FULL,
  block: DropStrategy.DROP_STRATEGY_BLOCK_IF_FULL,
  drop_newest: DropStrategy.DROP_STRATEGY_DROP_NEWEST,
  drop_oldest: DropStrategy.DROP_STRATEGY_DROP_OLDEST,
}

function dropStrategyToProto(value: string): DropStrategy {
  return DROP_STRATEGY_MAP[value.toLowerCase()] ?? DropStrategy.DROP_STRATEGY_UNSPECIFIED
}

// ---------------------------------------------------------------------------
// Converters
// ---------------------------------------------------------------------------

export const pluginMetadataToProto = (meta: SdkPluginMetadata): ProtoPluginMetadata => {
  return {
    name: meta.name,
    description: meta.description ?? '',
    schema: JSON.stringify(meta.schema ?? {}),
    inputs: (meta.inputs ?? []).map(propertyToProto),
    outputs: (meta.outputs ?? []).map(propertyToProto),
    ports: (meta.ports ?? []).map(portMetadataToProto),
    multiTrack: meta.multiTrack ? multiTrackConfigToProto(meta.multiTrack) : undefined,
  }
}

function propertyToProto(prop: PropertyDef): ProtoProperty {
  return {
    type: eventTypeToProto(prop.type),
    name: prop.name ?? '',
    fields: (prop.fields ?? []).map(fieldToProto),
  }
}

function fieldToProto(field: FieldDef): ProtoField {
  return {
    key: field.key,
    type: valueTypeToProto(field.type),
    required: field.required ?? false,
  }
}

function portMetadataToProto(port: SdkPortMetadata): ProtoPortMetadata {
  return {
    type: eventTypeToProto(port.type),
    port: port.port,
    name: port.name ?? '',
    description: port.description ?? '',
  }
}

function multiTrackConfigToProto(config: SdkMultiTrackConfig): ProtoMultiTrackConfig {
  return {
    enabled: config.enabled ?? true,
    payload: config.payload ? trackConfigToProto(config.payload) : undefined,
  }
}

function trackConfigToProto(config: SdkTrackConfig): ProtoTrackConfig {
  return {
    enabled: config.enabled ?? true,
    bufferSize: config.bufferSize ?? 128,
    dropStrategy: dropStrategyToProto(config.dropStrategy ?? 'block_if_full'),
    interruptSignals: config.interruptSignals ?? [],
  }
}
